File: internal/secrets/secrets.go
// Package secrets encrypts and decrypts values in exactly the format of the
// app's src/lib/services/secrets.ts: AES-256-GCM keyed by
// scrypt(AUTH_SECRET, "homerun-registry-secrets"), stored as
// base64(iv) + "." + base64(tag) + "." + base64(ciphertext).
package secrets

import (
	"crypto/aes"
	"crypto/cipher"
	"crypto/rand"
	"encoding/base64"
	"errors"
	"os"
	"strings"

	"golang.org/x/crypto/scrypt"
)

const (
	salt      = "homerun-registry-secrets"
	ivLength  = 12
	tagLength = 16
)

// ErrMalformed is returned for a stored value that isn't three base64 parts.
var ErrMalformed = errors.New("secrets: malformed ciphertext")

// Box encrypts and decrypts with one derived key.
type Box struct {
	aead cipher.AEAD
}

// AuthSecretFromEnv resolves the auth secret the same way the app's config
// does: AUTH_SECRET, else BETTER_AUTH_SECRET, else the "default-secret"
// placeholder.
func AuthSecretFromEnv() string {
	for _, key := range []string{"AUTH_SECRET", "BETTER_AUTH_SECRET"} {
		if value := os.Getenv(key); strings.TrimSpace(value) != "" {
			return value
		}
	}
	return "default-secret"
}

// New derives the key from authSecret with Node's scryptSync defaults
// (N=16384, r=8, p=1).
func New(authSecret string) (*Box, error) {
	key, err := scrypt.Key([]byte(authSecret), []byte(salt), 16384, 8, 1, 32)
	if err != nil {
		return nil, err
	}
	block, err := aes.NewCipher(key)
	if err != nil {
		return nil, err
	}
	aead, err := cipher.NewGCMWithNonceSize(block, ivLength)
	if err != nil {
		return nil, err
	}
	return &Box{aead: aead}, nil
}

// Encrypt seals plaintext with a fresh random IV.
func (b *Box) Encrypt(plaintext string) (string, error) {
	iv := make([]byte, ivLength)
	if _, err := rand.Read(iv); err != nil {
		return "", err
	}
	sealed := b.aead.Seal(nil, iv, []byte(plaintext), nil)
	ciphertext, tag := sealed[:len(sealed)-tagLength], sealed[len(sealed)-tagLength:]
	enc := base64.StdEncoding.EncodeToString
	return enc(iv) + "." + enc(tag) + "." + enc(ciphertext), nil
}

// Decrypt reverses Encrypt, or the TS encryptSecret().
func (b *Box) Decrypt(stored string) (string, error) {
	parts := strings.Split(stored, ".")
	if len(parts) != 3 {
		return "", ErrMalformed
	}
	decoded := make([][]byte, 3)
	for i, part := range parts {
		value, err := base64.StdEncoding.DecodeString(part)
		if err != nil {
			return "", ErrMalformed
		}
		decoded[i] = value
	}
	iv, tag, ciphertext := decoded[0], decoded[1], decoded[2]
	if len(iv) != ivLength || len(tag) != tagLength {
		return "", ErrMalformed
	}
	plaintext, err := b.aead.Open(nil, iv, append(ciphertext, tag...), nil)
	if err != nil {
		return "", err
	}
	return string(plaintext), nil
}

File: src/lib/services/secrets.ts
import {
	createCipheriv,
	createDecipheriv,
	randomBytes,
	scryptSync,
} from "node:crypto";
import { config } from "$lib/config";

// Registry credentials (service.registryPasswordEnc) are encrypted at
// rest with AES-256-GCM. The key is derived from the app's own auth
// secret via scrypt : no separate secret to manage/rotate for v1.
const ALGORITHM = "aes-256-gcm";
const IV_LENGTH = 12;

const cachedKeys = new Map<string, Buffer>();

function getKey(secret: string): Buffer {
	let key = cachedKeys.get(secret);
	if (!key) {
		key = scryptSync(secret, "homerun-registry-secrets", 32);
		cachedKeys.set(secret, key);
	}
	return key;
}

/**
 * Encrypts a plaintext into a single storable string, keyed by the app's auth
 * secret unless `secret` overrides it. internal/secrets (Go) reads and writes
 * the same format.
 */
export function encryptSecret(
	plaintext: string,
	secret: string = config.auth.secret,
): string {
	const iv = randomBytes(IV_LENGTH);
	const cipher = createCipheriv(ALGORITHM, getKey(secret), iv);
	const ciphertext = Buffer.concat([
		cipher.update(plaintext, "utf8"),
		cipher.final(),
	]);
	const authTag = cipher.getAuthTag();

	return [iv, authTag, ciphertext]
		.map((buf) => buf.toString("base64"))
		.join(".");
}

/** Reverses encryptSecret(). Returns null if the value can't be decrypted. */
export function decryptSecret(
	stored: string,
	secret: string = config.auth.secret,
): string | null {
	const parts = stored.split(".");
	if (parts.length !== 3) {
		return null;
	}
	const [ivB64, authTagB64, ciphertextB64] = parts;

	try {
		const iv = Buffer.from(ivB64, "base64");
		const authTag = Buffer.from(authTagB64, "base64");
		const ciphertext = Buffer.from(ciphertextB64, "base64");

		const decipher = createDecipheriv(ALGORITHM, getKey(secret), iv);
		decipher.setAuthTag(authTag);
		const plaintext = Buffer.concat([
			decipher.update(ciphertext),
			decipher.final(),
		]);
		return plaintext.toString("utf8");
	} catch {
		return null;
	}
}
